### PR TITLE
js: Add Jest testing

### PR DESCRIPTION
**Pull request by ix5**
As discussed in https://github.com/posativ/isso/issues/754, add some basic Jest testing config and examples to facilitate finally writing some frontend code tests.

---

**Note:** End-to-end tests using `puppeteer` were split off and moved into https://github.com/posativ/isso/pull/807

---

### Terminology
- **[Jest](https://jestjs.io/)** is a Javascript testing framework. It is mainly used here for unit tests.

### Running

Install the needed packages:
```sh
npm install
```

Then run:
- `npm run test-unit` for Jest unit tests

---

Highly **WIP** at the moment, but could be useful to someone else working on the same issues. Please communicate and collaborate!

Depends on https://github.com/posativ/isso/pull/800. ~~I have temporarily changed the base to that branch to make it easier to review this change.~~

### CODE REVIEW

**Comment by ix5** [comment]
Also pinging @stefangehn, @blatinier and @Lucas-C here. Let me know if you find this useful and please share tips/improvements/general ideas.

I'd love for someone more knowledgeable than me to write some proper tests and I hope this is a good starting-off point.

File: isso/js/jest-puppeteer.config.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
```
Comment (Lucas-C) [inline isso/js/jest-puppeteer.config.js:1]: This file should be at the root of the repository.
It is currently ignored by Jest when running `npm test`

File: isso/js/tests/unit/template.test.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// TODO not working because jest and pug-loader somehow don't want to work
```
Comment (Lucas-C) [inline isso/js/tests/unit/template.test.js:1]: Do you have the same error as me?
```
$ npm run test-unit

> test-unit
> jest --config isso/js/jest.config.js isso/js/tests/unit/

 PASS  isso/js/tests/unit/utils.test.js
 PASS  isso/js/tests/unit/config.test.js
 FAIL  isso/js/tests/unit/template.test.js
  ● Template functions

    ReferenceError: nodeResolve is not defined

      at Object.fn [as postbox] (eval at <anonymous> (app/template.js:28:13), <anonymous>:6:21)
      at Object.render (app/template.js:81:10)
      at Object.<anonymous> (tests/unit/template.test.js:13:19)
```

**Review by Lucas-C — COMMENTED** [review]
Nice bootstrapping! Good job @ix5 :)
Can I commit stuff on this branch if I find the time to write some tests?

**Comment by ix5** [comment]
Hey @Lucas-C, thank you so much for reviewing and verifying!

I probably should have split this PR into two PRs, one for Jest and one for the addition of `puppeteer`.

Re: `virtualenv` standard location and bringing up the Isso dev server, my wish for that would be to have some kind of standardization in testing.
Either docker containers with `docker-compose` for client and server (too heavyweight imo plus dependency on docker+compose) or an entry in the `Makefile` for creating the `virtualenv` and then starting the server from inside Node? Or maybe requiring the user to start the server?

Instrumentation is hard, and I'm open to suggestions here as I have no experience and no idea what accepted standard practices are.

File: isso/js/jest-puppeteer.config.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  server: {
+    command: '.venv-isso/bin/isso -c share/isso-dev.conf',
```
[the remark this answers is not in the document]
Comment (ix5) [reply isso/js/jest-puppeteer.config.js:3]: This config was more for myself or "for later use", but you're right, the location of the `virtualenv` should be standardized.

IMO we can drop Windows as it is currently also not in our GitHub actions CI testing matrices.

File: isso/js/tests/integration/puppet.test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//import 'expect-puppeteer';
+//const pup = require("expect-puppeteer");
+
+describe('isso', () => {
+  beforeAll(async () => {
+    //await page.goto('https://google.com')
+    await page.goto('http://localhost:8080/demo')
+  })
+
+  it('should display "Isso Demo" text on page', async () => {
```
[the remark this answers is not in the document]
Comment (ix5) [reply isso/js/tests/integration/puppet.test.js:10]: Yup, let's stick to that.

File: isso/js/tests/mocks/fileMock.js [hunk 5]
```diff
@@ -0,0 +1 @@
+module.exports = 'test-file-stub';
```
[the remark this answers is not in the document]
Comment (ix5) [reply isso/js/tests/mocks/fileMock.js:1]: It's for later on when `require`-ing `.svg` files (which outside of Jest, `webpack` takes care of).

File: package.json [hunk 6]
```diff
@@ -6,11 +6,15 @@
...
   "scripts": {
+    "test": "jest --config isso/js/jest.config.js isso/js/tests/" ,
+    "test-unit": "jest --config isso/js/jest.config.js isso/js/tests/unit/",
+    "test-integration": "jest --config isso/js/jest.config.js isso/js/tests/integration/",
     "build-dev": "webpack --config isso/js/webpack.config.js --config-name dev",
     "watch-dev": "webpack --config isso/js/webpack.config.js --config-name dev --watch",
     "build-prod": "webpack --config isso/js/webpack.config.js --merge --config-name dev --config-name prod"
   },
   "devDependencies": {
+    "jest": "^27.5.0",
```
[the remark this answers is not in the document]
Comment (ix5) [reply package.json:17]: Adding `jest-puppeteer` to `devDependencies` will automatically pull in a 400Mb+ download of headless chrome, as explained earlier. That's why I'm a bit hesitant to add it here.

Imo it should be up to the user whether they need to run the end-to-end tests.

Maybe there's an option to configure the package _before_ installing to not pull in the chrome download automatically?

File: isso/js/jest-puppeteer.config.js (hunk 1 above)
Comment (ix5) [reply isso/js/jest-puppeteer.config.js:1]: I'd like to keep all javascript-related stuff inside `isso/js/`, so I'll have to figure out a way for Jest to read this file.
See https://github.com/smooth-code/jest-puppeteer#configure-puppeteer

(remains TODO)

File: isso/js/tests/unit/template.test.js (hunk 2 above)
Comment (ix5) [reply isso/js/tests/unit/template.test.js:1]: Yeah `pug-loader` and Jest will not work together out of the box. This test is currently expected to fail.

**Comment by ix5** [comment]
> Nice bootstrapping! Good job @ix5 :) Can I commit stuff on this branch if I find the time to write some tests?

Let me split up the `puppeteer` addition into another PR first, then commit away ;)

File: isso/js/tests/unit/utils.test.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+const utils = require("app/utils");
+
+describe("Padding function", () => {
```
[the remark this answers is not in the document]
Comment (ix5) [reply isso/js/tests/unit/utils.test.js]: Fixed

File: isso/js/tests/unit/template.test.js (hunk 2 above)
Comment (ix5) [reply isso/js/tests/unit/template.test.js]: Disabled this test for onw

File: isso/js/jest-puppeteer.config.js (hunk 1 above)
Comment (ix5) [reply isso/js/jest-puppeteer.config.js]: Point of reference: https://github.com/smooth-code/jest-puppeteer/issues/160#issuecomment-491975158
```js
process.env.JEST_PUPPETEER_CONFIG = require.resolve('./jest-puppeteer.config.js');
```

**Pull request merged**